scalafix-core/shared/src/main/scala/scalafix/internal/util/TypeSyntax.scala

```diff
@@ -1,138 +1,451 @@
 import scala.reflect.ClassTag
+import scala.util.control.NonFatal
 import scalafix._
 import scalafix.internal.util.SymbolOps.SymbolType
-import scalafix.util.SymbolMatcher
-import scalafix.util.TreeExtractors._
 import org.langmeta.semanticdb.Symbol
+import scala.meta.internal.semanticdb3.SingletonType.{Tag => x}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Property => p}
+import scala.meta.internal.semanticdb3.Type.{Tag => t}
+import TypeSyntax._
+import scalapb.GeneratedMessage
 
 object TypeSyntax {
-  def prettify(tpe: Type, ctx: RuleCtx, shortenNames: Boolean)(
-      implicit index: SemanticdbIndex): (Type, Patch) = {
+  def prettify(
+      stpe: s.Type,
+      ctx: RuleCtx,
+      shortenNames: Boolean,
+      pos: Position)(implicit index: SemanticdbIndex): Option[(Type, Patch)] = {
+    try {
+      val table = index.asInstanceOf[EagerInMemorySemanticdbIndex].table
+      Some(new TypeSyntax(stpe, ctx, shortenNames, pos, table).prettify())
+    } catch {
+      case NonFatal(e) =>
+        val msg = new StringWriter()
+        e.printStackTrace(new PrintWriter(msg))
+        ctx.config.reporter.error(msg.toString, pos)
+        None
+    }
+  }
 
-    val functionN: SymbolMatcher = SymbolMatcher.exact(
-      0.to(22).map(i => Symbol(s"_root_.scala.Function$i#")): _*
-    )
-    val tupleN: SymbolMatcher = SymbolMatcher.exact(
-      1.to(22).map(i => Symbol(s"_root_.scala.Tuple$i#")): _*
-    )
+  implicit class XtensionSymbolInformationProperties(
+      info: s.SymbolInformation) {
+    private def test(property: s.SymbolInformation.Property): Boolean =
+      (info.properties & property.value) != 0
+    def isVal: Boolean = test(p.VAL)
+    def isVar: Boolean = test(p.VAR)
+    def isVarSetter: Boolean =
+      isVar && info.name.endsWith("_=")
+  }
 
-    def isStable(symbol: Symbol): Boolean = {
-      def loop(symbol: Symbol): Boolean = symbol match {
-        case Symbol.None => true
-        case Symbol.Global(owner, Signature.Term(_)) => loop(owner)
-        case _ => false
-      }
-      symbol match {
-        case Symbol.Global(owner, Signature.Term(_) | Signature.Type(_)) =>
-          loop(owner)
+  object T {
+    abstract class TypeRefExtractor(sym: String) {
+      def matches(tpe: s.Type): Boolean = tpe.tag match {
+        case t.TYPE_REF =>
+          tpe.typeRef.exists(_.symbol == sym)
+        case t.WITH_TYPE =>
+          tpe.withType.exists { x =>
+            x.types.lengthCompare(1) == 0 &&
+            unapply(x.types.head)
+          }
         case _ =>
           false
       }
+      def unapply(tpe: s.Type): Boolean = matches(tpe)
+    }
+    object AnyRef extends TypeRefExtractor("scala.AnyRef#")
+    object Nothing extends TypeRefExtractor("scala.Nothing#")
+    object Any extends TypeRefExtractor("scala.Any#")
+  }
+
+  def isFunctionN(symbol: String): Boolean = {
+    symbol.startsWith("scala.Function") &&
```

Comment:
I don't know if there this is a perforamance issue, but from an accuracy issue how about "scala.Function(\\d)+#".r to make sure we capture Function1 but not FunctionPure (not that such a thing currently exists).

Reply:
there are no other `scala.Function*` symbols so I thought this would be enough, regexes are such a heavy machine.